src/cockpit/samples.py

```diff
@@ -192,49 +192,54 @@ class CGroupSampler(Sampler):
     cgroups_v2: Optional[bool] = None
     cgroups_v2_path = '/sys/fs/cgroup/'
 
-    def read_cgroup_keyed_stat(self, samples, path, cgroup, name, key):
-        with open(path) as stat:
-            for line in stat:
-                if not line.startswith(key):
-                    continue
-
-                value = int(line.split()[-1])
-                if sys.maxsize > value > 0:
-                    samples[name][cgroup] = value / 1000
-
-    def read_cgroup_integer_stat(self, samples, path, cgroup, name):
+    @staticmethod
+    def read_cgroup_integer_stat(rootfd, statfile, include_zero=False, key=b''):
         # Not every stat is available, such as cpu.weight
         try:
-            with open(path) as stat:
-                # Some samples such as "memory.max" contains "max" when there is a no limit
-                try:
-                    value = int(stat.read().strip())
-                except ValueError:
-                    return
-
-                if sys.maxsize > value > 0:
-                    samples[name][cgroup] = value
+            fd = os.open(statfile, os.O_RDONLY, dir_fd=rootfd)
         except FileNotFoundError:
-            pass
+            return None
+
+        try:
+            data = os.read(fd, 1024)
+        finally:
+            os.close(fd)
+
+        if key:
+            start = data.index(key) + len(key)
+            end = data.index(b'\n', start)
+            # cpu.stat are in usecs
+            value = int(data[start:end]) / 1000
```

Comment:
I don't like how this `/ 1000` is hardcoded.  It makes the `key` path function fundamentally differently from the unkeyed path, in an unexpected way.  Please move the `/ 1000` to the call site before assigning the value to the dictionary.  Unless: is there a chance that the result might be None?  I wonder if it might make sense here to do NaN.

Reply:
The result can be None for sure, so I'll break it up and add a check.